lumicks/pylake/detail/imaging_mixins.py

```diff
@@ -1,11 +1,131 @@
+import itertools
+import json
+import numpy as np
+import numpy.typing as npt
+import tifffile
+from typing import Iterator, Union
+
 from ..adjustments import ColorAdjustment
 from .timeindex import to_timestamp
-import numpy as np
-from typing import Union
 
 _FIRST_TIMESTAMP = 1388534400
 
 
+class TiffExport:
+    def export_tiff(self, filename, *, dtype=None, clip=False):
+        """Export frames to a TIFF image
+
+        Parameters
+        ----------
+        filename : str
+            The name of the TIFF file where the image will be saved.
+        dtype : np.dtype
+            The data type of a single color channel in the resulting image.
+        clip : bool
+            If enabled, the image data will be clipped to fit into the desired `dtype`. This option
+            is disabled by default: an error will be raise if the data does not fit.
+        """
+        # If the the exported tiff should be casted to `dtype`, get the full image stack to later
+        # safely cast it. Otherwise, try to get an iterator, to save memory.
+        frames = self._tiff_frames(iterator=dtype is None)
+
+        def _check_empty(frames):
+            if hasattr(frames, "__next__"):
+                # `frames` is an iterator. Peek into the first frame to check if there are any.
+                def peek(iterable):
+                    iterator = iterable
+                    try:
+                        first = next(iterator)
+                    except StopIteration:
+                        return None
+                    return itertools.chain([first], iterator)
+
+                frames = peek(frames)
+            else:
+                # `frames` is an np.ndarray. Check the size.
+                frames = frames if frames.size > 0 else None
+
+            if frames is None:
+                raise RuntimeError("Can't export TIFF if there are no pixels")
+
+            return frames
+
+        frames = _check_empty(frames)
+
+        # Cast frames if a specific dtype is requested.
+        if dtype:
+
+            def _cast_image(image, dtype=np.float32, clip=False):
+                # Check if requested dtype can fit image values without an overflow
+                info = np.finfo(dtype) if np.dtype(dtype).kind == "f" else np.iinfo(dtype)
+                if not clip and (np.min(image) < info.min or np.max(image) > info.max):
+                    raise RuntimeError(
+                        f"Can't safely export image with `dtype={dtype.__name__}` channels."
+                        f" Switch to a larger `dtype` in order to safely store everything"
+                        f" or pass `force=True` to clip the data."
+                    )
+
+                return image.astype(dtype)
+
+            frames = _cast_image(frames, dtype=dtype, clip=clip)
```

Comment:
Personally, I don't like putting the entire function indented here and doing something like this (but that's more a personal preference):

```python
        # Cast frames if a specific dtype is requested.
        def _cast_image(image, dtype=np.float32, clip=False):
            # Check if requested dtype can fit image values without an overflow
            info = np.finfo(dtype) if np.dtype(dtype).kind == "f" else np.iinfo(dtype)
            if not clip and (np.min(image) < info.min or np.max(image) > info.max):
                raise RuntimeError(
                    f"Can't safely export image with `dtype={dtype.__name__}` channels."
                    f" Switch to a larger `dtype` in order to safely store everything"
                    f" or pass `force=True` to clip the data."
                )

            return image.astype(dtype)

        if dtype:
            frames = _cast_image(frames, dtype=dtype, clip=clip)
```

Or even:
```python
   frames = _cast_image(frames, dtype=dtype, clip=clip) if dtype else frames
```

Reply:
OK, I agree it is nicer to reduce indentations.

Changed.